File: server/internal/models/rule.go
package models

import (
	"time"
)

type RuleFieldType string
type RuleOperator string
type ConditionLogic string

const (
	RuleFieldAmount      RuleFieldType = "amount"
	RuleFieldName        RuleFieldType = "name"
	RuleFieldDescription RuleFieldType = "description"
	RuleFieldCategory    RuleFieldType = "category"
	RuleFieldTransfer    RuleFieldType = "transfer"
)

const (
	OperatorEquals   RuleOperator = "equals"
	OperatorContains RuleOperator = "contains"
	OperatorGreater  RuleOperator = "greater"
	OperatorLower    RuleOperator = "lower"
)

const (
	ConditionLogicAnd ConditionLogic = "AND"
	ConditionLogicOr  ConditionLogic = "OR"
)

type CreateBaseRuleRequest struct {
	Name           string          `json:"name" binding:"required,min=1,max=100"`
	Description    *string         `json:"description,omitempty" binding:"omitempty,max=255"`
	ConditionLogic *ConditionLogic `json:"condition_logic" binding:"omitempty,oneof=AND OR"`
	EffectiveFrom  time.Time       `json:"effective_from" binding:"required"`
	CreatedBy      int64           `json:"created_by"`
}

type CreateRuleRequest struct {
	Rule       CreateBaseRuleRequest        `json:"rule" binding:"required"`
	Actions    []CreateRuleActionRequest    `json:"actions" binding:"required,min=1"`
	Conditions []CreateRuleConditionRequest `json:"conditions" binding:"required,min=1"`
}

type UpdateRuleRequest struct {
	Name           *string         `json:"name,omitempty" binding:"omitempty,max=100"`
	Description    *string         `json:"description,omitempty" binding:"omitempty,max=255"`
	ConditionLogic *ConditionLogic `json:"condition_logic" binding:"omitempty,oneof=AND OR"`
	EffectiveFrom  *time.Time      `json:"effective_from,omitempty"`
}

type RuleResponse struct {
	Id             int64          `json:"id"`
	Name           string         `json:"name"`
	Description    *string        `json:"description"`
	ConditionLogic ConditionLogic `json:"condition_logic"`
	EffectiveFrom  time.Time      `json:"effective_from"`
	CreatedBy      int64          `json:"created_by"`
}

type DescribeRuleResponse struct {
	Rule       RuleResponse            `json:"rule"`
	Actions    []RuleActionResponse    `json:"actions"`
	Conditions []RuleConditionResponse `json:"conditions"`
}

type CreateRuleActionRequest struct {
	ActionType  RuleFieldType `json:"action_type" binding:"required"`
	ActionValue string        `json:"action_value" binding:"required,min=1,max=100"`
	RuleId      int64         `json:"rule_id"`
}

type UpdateRuleActionRequest struct {
	ActionType  *RuleFieldType `json:"action_type"`
	ActionValue *string        `json:"action_value"`
}

type RuleActionResponse struct {
	Id          int64         `json:"id"`
	RuleId      int64         `json:"rule_id"`
	ActionType  RuleFieldType `json:"action_type"`
	ActionValue string        `json:"action_value"`
}

type CreateRuleConditionRequest struct {
	ConditionType     RuleFieldType `json:"condition_type" binding:"required"`
	ConditionValue    string        `json:"condition_value" binding:"required,min=1,max=100"`
	ConditionOperator RuleOperator  `json:"condition_operator" binding:"required"`
	RuleId            int64         `json:"rule_id"`
}

type UpdateRuleConditionRequest struct {
	ConditionType     *RuleFieldType `json:"condition_type"`
	ConditionValue    *string        `json:"condition_value"`
	ConditionOperator *RuleOperator  `json:"condition_operator"`
}

type RuleConditionResponse struct {
	Id                int64         `json:"id"`
	RuleId            int64         `json:"rule_id"`
	ConditionType     RuleFieldType `json:"condition_type"`
	ConditionValue    string        `json:"condition_value"`
	ConditionOperator RuleOperator  `json:"condition_operator"`
}

type PutRuleActionsRequest struct {
	Actions []CreateRuleActionRequest `json:"actions" binding:"required,min=1,max=50"`
}

type PutRuleConditionsRequest struct {
	Conditions []CreateRuleConditionRequest `json:"conditions" binding:"required,min=1,max=50"`
}

type PutRuleActionsResponse struct {
	Actions []RuleActionResponse `json:"actions"`
}

type PutRuleConditionsResponse struct {
	Conditions []RuleConditionResponse `json:"conditions"`
}

type ExecuteRulesRequest struct {
	RuleIds        *[]int64 `json:"rule_ids,omitempty"`
	TransactionIds *[]int64 `json:"transaction_ids,omitempty"`
	PageSize       int      `json:"page_size,omitempty"`
}

type ExecuteRulesResponse struct {
	Modified      []ModifiedResult `json:"modified"`
	Skipped       []SkippedResult  `json:"skipped"`
	TotalRules    int              `json:"total_rules"`
	ProcessedTxns int              `json:"processed_transactions"`
}

type ModifiedResult struct {
	TransactionId int64           `json:"transaction_id"`
	AppliedRules  []int64         `json:"applied_rules"`
	UpdatedFields []RuleFieldType `json:"updated_fields"`
}

type SkippedResult struct {
	TransactionId int64  `json:"transaction_id"`
	Reason        string `json:"reason"`
}

// PaginatedRulesResponse is the paginated response for rule listing
type PaginatedRulesResponse struct {
	Rules    []RuleResponse `json:"rules"`
	Total    int            `json:"total"`
	Page     int            `json:"page"`
	PageSize int            `json:"page_size"`
}

// RuleListQuery holds query params for filtering, sorting, and pagination
type RuleListQuery struct {
	Page     int     // page number (1-based)
	PageSize int     // items per page
	Search   *string // search in name/description
}

File: frontend/lib/models/rule.ts
export enum ConditionLogic {
  AND = "AND",
  OR = "OR",
}

export type RuleFieldType = "amount" | "name" | "description" | "category" | "transfer";
export type RuleOperator = "equals" | "contains" | "greater" | "lower";

export const RULE_FIELD_TYPES: { label: string; value: RuleFieldType }[] = [
  { label: "Amount", value: "amount" },
  { label: "Name", value: "name" },
  { label: "Description", value: "description" },
  { label: "Category", value: "category" },
  { label: "Transfer", value: "transfer" },
];

export const RULE_OPERATORS: { label: string; value: RuleOperator }[] = [
  { label: "Equals", value: "equals" },
  { label: "Contains", value: "contains" },
  { label: "Greater", value: "greater" },
  { label: "Lower", value: "lower" },
];

export interface BaseRule {
  name: string;
  description?: string;
  condition_logic?: ConditionLogic;
  effective_from: string;
}

export interface BaseRuleAction {
  action_type: RuleFieldType;
  action_value: string;
}

export interface BaseRuleCondition {
  condition_type: RuleFieldType;
  condition_value: string;
  condition_operator: RuleOperator;
}

export type CreateBaseRuleInput = BaseRule;
export type CreateRuleActionInput = BaseRuleAction;
export type CreateRuleConditionInput = BaseRuleCondition;

export interface CreateRuleInput {
  rule: CreateBaseRuleInput;
  actions: CreateRuleActionInput[];
  conditions: CreateRuleConditionInput[];
}

export type UpdateRuleInput = Partial<CreateBaseRuleInput>;
export type UpdateRuleActionInput = Partial<CreateRuleActionInput>;
export type UpdateRuleConditionInput = Partial<CreateRuleConditionInput>;

export interface Rule extends BaseRule {
  id: number;
  created_by: number;
  condition_logic: ConditionLogic;
}

export interface RuleAction extends BaseRuleAction {
  id: number;
  rule_id: number;
}

export interface RuleCondition extends BaseRuleCondition {
  id: number;
  rule_id: number;
}

export interface DescribeRuleResponse {
  rule: Rule;
  actions: RuleAction[];
  conditions: RuleCondition[];
}

export interface ExecuteRulesResponse {
  modified: ModifiedResult[];
  processed_transactions: SkippedResult[];
}

export interface ModifiedResult {
  transaction_id: number;
  applied_rules: number[];
  updated_fields: RuleFieldType[];
}

export interface SkippedResult {
  transaction_id: number;
  reason: string;
}
